### DISCUSSION TITLE
Unable to import jax

### DISCUSSION THREAD

**Discussion by gvijay1788** [jax-ml/jax#13988] [General]
Hello, I tried isntalling different version of jax and jaxlib and getting following error message while importing jax (import jax). Can anyone assist in this.

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

jax | Jaxlib | errror
-- | -- | --
0.3.13 | 0.3.5 | AttributeError: module 'jax' has   no attribute 'version'
jax-0.3.25 |   | ImportError: cannot import name   'custom_ivjp' from 'jax.api'   (/home/notebook/.virtualenvs/1267/lib/python3.7/site-packages/jax/api.py)
jax 0.1.75 | jaxlib 0.3.5 | AttributeError: 'str' object has   no attribute '__name__'

</div>

</body>

</html>

I am sharing screenshot of one of error.  I have tried all different suggestion on google, unable to fix it. kindly suggest solution.

![image](https://user-images.githubusercontent.com/57652340/212165971-00fd43ba-45a9-4230-933e-70e3990a9042.png)

**Comment by jakevdp** [comment]
Is there a reason you prefer older JAX versions?

For JAX v0.3.0 or newer, the rule of thumb is to use the most recent jaxlib release that is not newer than the jax release. So JAX 0.3.15 goes with jaxlib 0.3.15, and jax 0.3.25 goes with jaxlib 0.3.25, and jax 0.3.19 goes with jaxlib 0.3.15 (because the next jaxlib release after 0.3.15 was 0.3.20). You can find more on this policy here: https://jax.readthedocs.io/en/latest/jep/9419-jax-versioning.html

If you need JAX releases older than 0.3.0, I would recommend first figuring out why that's required and trying to use a newer release instead. But if you still absolutely need an older jax version, I'd recommend finding the most recent jaxlib at the time of its release: the release history is listed at https://pypi.org/project/jax/#history and https://pypi.org/project/jaxlib/#history). For example, jax 0.2.10 was released on March 5 2021, at which point the most recent jaxlib release was 0.1.61 and so those versions should work together.

**Reply by gvijay1788** [reply]
Thanks jakevdp, I tried your suggstions but still getting err0r-

**1.** pip install jax==0.3.15 jaxlib==0.3.15
          **Error:**  import jax result following error-
         AttributeError: module 'jax' has no attribute 'version'

**2.** pip install jax==0.3.25 jaxlib==0.3.25
          **Error:** AttributeError: module 'jax' has no attribute 'version'

**3.** pip install jax==0.3.19 jaxlib==0.3.15
          **Error:** AttributeError: module 'jax' has no attribute 'version'

![image](https://user-images.githubusercontent.com/57652340/212169887-cda0d52c-4a17-4e16-8cb5-3ef15ed53cdf.png)

**Reply by gvijay1788** [reply]
I am happy if newer version works well.

**Reply by jakevdp** [reply]
Thanks for the details. the error you're getting is strange, because `jax.version` is explicitly imported in the file above the line where this error is being raised (see https://github.com/google/jax/blob/jaxlib-v0.3.15/jax/_src/lib/__init__.py#L45). This makes me think that Python's import machinery is finding multiple copies of jax.

Do you by chance have multiple versions of `jax` in your Python import path? Or perhaps do you have a file named `jax.py` or a directory named `jax` in your path? I would start by digging in there, rather than trying to install different jax versions.

**Reply by jakevdp** [reply]
Actually, I bet this is the issue: if you're running in a jupyter or colab notebook, you need to restart your runtime after installing a new version of the package, otherwise the previously-imported version will be cached in `sys.modules` and you can encounter strange errors like this.

You can see this recommendation in the text in your screenshot, though it's easy to miss.

**Comment by gvijay1788** [comment]
Thanks for your suggestion. I am able to install with different version of python. Now M running following code and getting error while adding assertion.
Can you please help in this.

```python
from jax.lib import xla_bridge
assert xla_bridge.get_backend().platform == 'gpu', 'Failed to locate GPU'
```

![image](https://user-images.githubusercontent.com/57652340/212206240-74758d5c-7aca-4238-b451-0f33101f86b2.png)

Can you please suggest.

**Reply by gvijay1788** [reply]
I also see default backend as cpu. How can I change it ? 

![image](https://user-images.githubusercontent.com/57652340/212206470-a5a0e9b3-2387-400e-a5e3-aee441a35a15.png)

**Reply by jakevdp** [reply]
If you want to use GPU with JAX, you need to install a GPU-compatible jaxlib build. See the [GPU Install Guide](https://github.com/google/jax#pip-installation-gpu-cuda) for more information.

**Reply by gvijay1788** [reply]
Thanks for your guide, I installed GPU but this is having different error now -

![image](https://user-images.githubusercontent.com/57652340/212210946-7e57660d-b3e0-4463-9c2b-25b6f7beeb50.png)

**Reply by gvijay1788** [reply]
Earlier I had following version of jax and jaxlib -
pip install jax==0.3.25 jaxlib==0.3.25

Now I tried with newer version of jax and jaxlib, throwing error-

pip install jax==0.3.15 jaxlib==0.3.15

![image](https://user-images.githubusercontent.com/57652340/212212785-26f84ec4-fd1a-409d-8288-a4bb5e163e7c.png)

**Reply by jakevdp** [reply]
All of these import errors look like version skew between jax and jaxlib. Make sure you are installing compatible jax & jaxlib versions, and then make sure you restart your runtime to clear the `sys.modules` cache of the old versions.

**Reply by gvijay1788** [reply]
I have installed matching versions what is suggested above. either both 0.3.15 or 0.3.25 version.
After running sys.module.clear(), code stuck.

![image](https://user-images.githubusercontent.com/57652340/212216569-f7f12b52-94d7-4f8c-86b1-4e03e38d0293.png)

**Reply by jakevdp** [reply]
I would not rely on `sys.modules.clear()`, I would actually restart the runtime. Everything you're running into looks like a symptom of not restarting the runtime after installing a new package version.

**Reply by gvijay1788** [reply]
I restarted kernal as well couple of times, its unable to locate GPU.

![image](https://user-images.githubusercontent.com/57652340/212224506-d814d4ea-4c28-4a1b-95fe-01e4295fb283.png)

Can you please suggest complete installation with GPU steps if you have  ?

**Reply by jakevdp** [reply]
It’s hard to say what’s going on without more information. I’d suggest following the steps I linked to above, ensuring that your CUDA version is correct, and then open an issue if you run into problems.

**Reply by gvijay1788** [reply]
What Cuda version needed for following combinations: 
 jax==0.3.25 jaxlib==0.3.25

**Reply by jakevdp** [reply]
The easiest way to find this information for any particular jaxlib version is by searching the list at https://storage.googleapis.com/jax-releases/ The targeted cuda and cudnn versions are listed as part of the filename. For example, `cuda11/jaxlib-0.3.25+cuda11.cudnn82-cp38-cp38-manylinux2014_x86_64.whl` is jaxlib version 0.3.25, built for CUDA 11.X and CUDNN 8.2, targeting Python 3.8 running on linux with x86_64 architecture.

**Reply by gvijay1788** [reply]
It worked, Thank you.

**Reply by gvijay1788** [reply]
Now I am trying to run beloe and getting error-
Code:
```
import jax.numpy as jnp
jnp.zeros(30)
```
Error:
```
---------------------------------------------------------------------------
XlaRuntimeError                           Traceback (most recent call last)
<ipython-input-8-47207b5c4048> in <module>
      1 import jax.numpy as jnp
----> 2 jnp.zeros(30)

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/numpy/lax_numpy.py in zeros(shape, dtype)
   2097   lax_internal._check_user_dtype_supported(dtype, "zeros")
   2098   shape = canonicalize_shape(shape)
-> 2099   return lax.full(shape, 0, _jnp_dtype(dtype))
   2100 
   2101 @_wraps(np.ones)

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/lax/lax.py in full(shape, fill_value, dtype)
   1197   dtype = dtypes.canonicalize_dtype(dtype or _dtype(fill_value))
   1198   fill_value = _convert_element_type(fill_value, dtype, weak_type)
-> 1199   return broadcast(fill_value, shape)
   1200 
   1201 def zeros_like_shaped_array(aval: ArrayLike) -> Array:

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/lax/lax.py in broadcast(operand, sizes)
    792   """
    793   dims = tuple(range(len(sizes), len(sizes) + np.ndim(operand)))
--> 794   return broadcast_in_dim(operand, tuple(sizes) + np.shape(operand), dims)
    795 
    796 def broadcast_in_dim(operand: ArrayLike, shape: Shape,

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/lax/lax.py in broadcast_in_dim(operand, shape, broadcast_dimensions)
    822   return broadcast_in_dim_p.bind(
    823       operand, *dyn_shape, shape=tuple(static_shape),
--> 824       broadcast_dimensions=tuple(broadcast_dimensions))
    825 
    826 def broadcast_to_rank(x: Array, rank: int) -> Array:

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/core.py in bind(self, *args, **params)
    327     assert (not config.jax_enable_checks or
    328             all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 329     return self.bind_with_trace(find_top_trace(args), args, params)
    330 
    331   def bind_with_trace(self, trace, args, params):

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/core.py in bind_with_trace(self, trace, args, params)
    330 
    331   def bind_with_trace(self, trace, args, params):
--> 332     out = trace.process_primitive(self, map(trace.full_raise, args), params)
    333     return map(full_lower, out) if self.multiple_results else full_lower(out)
    334 

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/core.py in process_primitive(self, primitive, tracers, params)
    710 
    711   def process_primitive(self, primitive, tracers, params):
--> 712     return primitive.impl(*tracers, **params)
    713 
    714   def process_call(self, primitive, f, tracers, params):

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in apply_primitive(prim, *args, **params)
    112   """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
    113   compiled_fun = xla_primitive_callable(prim, *unsafe_map(arg_spec, args),
--> 114                                         **params)
    115   return compiled_fun(*args)
    116 

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/util.py in wrapper(*args, **kwargs)
    251         return f(*args, **kwargs)
    252       else:
--> 253         return cached(config._trace_context(), *args, **kwargs)
    254 
    255     wrapper.cache_clear = cached.cache_clear

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/util.py in cached(_, *args, **kwargs)
    244     @functools.lru_cache(max_size)
    245     def cached(_, *args, **kwargs):
--> 246       return f(*args, **kwargs)
    247 
    248     @functools.wraps(f)

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in xla_primitive_callable(prim, *arg_specs, **params)
    196       return out,
    197   compiled = _xla_callable_uncached(lu.wrap_init(prim_fun), device, None,
--> 198                                     prim.name, donated_invars, False, *arg_specs)
    199   if not prim.multiple_results:
    200     return lambda *args, **kw: compiled(*args, **kw)[0]

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in _xla_callable_uncached(fun, device, backend, name, donated_invars, keep_unused, *arg_specs)
    358   else:
    359     return lower_xla_callable(fun, device, backend, name, donated_invars, False,
--> 360                               keep_unused, *arg_specs).compile().unsafe_call
    361 
    362 xla_callable = lu.cache(_xla_callable_uncached)

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in compile(self)
    996         self._executable = XlaCompiledComputation.from_xla_computation(
    997             self.name, self._hlo, self._in_type, self._out_type,
--> 998             **self.compile_args)
    999 
   1000     return self._executable

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in from_xla_computation(name, xla_computation, in_type, out_type, nreps, device, backend, tuple_args, in_avals, out_avals, has_unordered_effects, ordered_effects, kept_var_idx, keepalive, host_callbacks)
   1193                           "in {elapsed_time} sec"):
   1194       compiled = compile_or_get_cached(backend, xla_computation, options,
-> 1195                                        host_callbacks)
   1196     buffer_counts = get_buffer_counts(out_avals, ordered_effects,
   1197                                       has_unordered_effects)

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in compile_or_get_cached(backend, computation, compile_options, host_callbacks)
   1076 
   1077   return backend_compile(backend, serialized_computation, compile_options,
-> 1078                          host_callbacks)
   1079 
   1080 

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/profiler.py in wrapper(*args, **kwargs)
    312   def wrapper(*args, **kwargs):
    313     with TraceAnnotation(name, **decorator_kwargs):
--> 314       return func(*args, **kwargs)
    315     return wrapper
    316   return wrapper

~/.virtualenvs/tkt1277/lib/python3.7/site-packages/jax/_src/dispatch.py in backend_compile(backend, built_c, options, host_callbacks)
   1010   # TODO(sharadmv): remove this fallback when all backends allow `compile`
   1011   # to take in `host_callbacks`
-> 1012   return backend.compile(built_c, compile_options=options)
   1013 
   1014 # TODO(phawkins): update users.

XlaRuntimeError: FAILED_PRECONDITION: Couldn't get ptxas version string: INTERNAL: Running ptxas --version returned 32512
```

Can you please suggest.

**Reply by jakevdp** [reply]
This is generally related to your system being unable to find CUDA binaries. Make certain CUDA is in your `$PATH`; see https://github.com/google/jax/discussions/6843

**Reply by gvijay1788** [reply]
Instaled below- 
pip install --user nvidia-pyindex
pip install --user nvidia-cuda-nvcc

And I see nvidia_cuda_nvcc-11.3.58.dist-info in site-packages directory. This didn't resolve it.

**Reply by jakevdp** [reply]
I think I recall there still is some blocker to using pip-installed CUDA with JAX. @hawkinsp might know the details

**Reply by gvijay1788** [reply]
Hello @hawkinsp : Could you please assist in this ? ?

**Reply by gvijay1788** [reply]
@jakevdp : Can you please help provide help on this ?

**Reply by jakevdp** [reply]
I would suggest not using pip installed CUDA for now. I recall hearing that there is some problem with that in the current release, but I don’t recall specifically what it is.

**Reply by gvijay1788** [reply]
and did we figure out solution or alternate for this problem? Could you please help me out to know that? May be get it from [hawkinsp] is you working together.

**Reply by jakevdp** [reply]
For help with installing CUDA, I'd recommend following the CUDA docs; e.g. https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html

**Reply by jakevdp** [reply]
FYI, there's some info here on why pip-installed CUDA does not work with JAX: https://github.com/google/jax/pull/13008#issuecomment-1305994934 We're hoping NVIDIA is able to fix that soon so we can streamline the jaxlib-gpu install process.
